src/hotspot/share/utilities/hashtable.cpp

```diff
@@ -45,56 +44,21 @@
 #include "utilities/hashtable.inline.hpp"
 #include "utilities/numberSeq.hpp"
 
-
 // This hashtable is implemented as an open hash table with a fixed number of buckets.
 
-template <MEMFLAGS F> BasicHashtableEntry<F>* BasicHashtable<F>::new_entry_free_list() {
-  BasicHashtableEntry<F>* entry = NULL;
-  if (_free_list != NULL) {
-    entry = _free_list;
-    _free_list = _free_list->next();
-  }
-  return entry;
-}
+// Hashtable entry allocates in the C heap directly.
 
-// HashtableEntrys are allocated in blocks to reduce the space overhead.
 template <MEMFLAGS F> BasicHashtableEntry<F>* BasicHashtable<F>::new_entry(unsigned int hashValue) {
-  BasicHashtableEntry<F>* entry = new_entry_free_list();
-
-  if (entry == NULL) {
-    if (_first_free_entry + _entry_size >= _end_block) {
-      int block_size = MAX2((int)_table_size / 2, (int)_number_of_entries); // pick a reasonable value
-      block_size = clamp(block_size, 2, 512); // but never go out of this range
-      int len = round_down_power_of_2(_entry_size * block_size);
-      assert(len >= _entry_size, "");
-      _first_free_entry = NEW_C_HEAP_ARRAY2(char, len, F, CURRENT_PC);
-      _entry_blocks.append(_first_free_entry);
-      _end_block = _first_free_entry + len;
-    }
-    entry = (BasicHashtableEntry<F>*)_first_free_entry;
-    _first_free_entry += _entry_size;
-  }
-
+  BasicHashtableEntry<F>* entry = ::new (NEW_C_HEAP_ARRAY(char, this->entry_size(), F))
+                                        BasicHashtableEntry<F>(hashValue);
   assert(_entry_size % HeapWordSize == 0, "");
-  entry->set_hash(hashValue);
   return entry;
 }
 
 
 template <class T, MEMFLAGS F> HashtableEntry<T, F>* Hashtable<T, F>::new_entry(unsigned int hashValue, T obj) {
-  HashtableEntry<T, F>* entry;
-
-  entry = (HashtableEntry<T, F>*)BasicHashtable<F>::new_entry(hashValue);
-  entry->set_literal(obj);
-  return entry;
-}
-
-// Version of hashtable entry allocation that allocates in the C heap directly.
-// The block allocator in BasicHashtable has less fragmentation, but the memory is not freed until
-// the whole table is freed. Use allocate_new_entry() if you want to individually free the memory
-// used by each entry
-template <class T, MEMFLAGS F> HashtableEntry<T, F>* Hashtable<T, F>::allocate_new_entry(unsigned int hashValue, T obj) {
-  HashtableEntry<T, F>* entry = (HashtableEntry<T, F>*) NEW_C_HEAP_ARRAY(char, this->entry_size(), F);
+  HashtableEntry<T, F>* entry = ::new (NEW_C_HEAP_ARRAY(char, this->entry_size(), F))
+                                      HashtableEntry<T, F>(hashValue);
 
   if (DumpSharedSpaces) {
     // Avoid random bits in structure padding so we can have deterministic content in CDS archive
```

Comment:
Hmm, the sequence looks a little odd: the constructor initializes some fields, they are then zeroed here, and then initialized again below ....

Actually, I think you can get rid of the `if (DumpSharedSpaces)` block for now. I am not sure if it's needed. Deterministic CDS archive is broken anyway (https://bugs.openjdk.java.net/browse/JDK-8253495).

That way you can get rid of the entry->set_xxx below as well.

Reply:
Yes, I'd like to make the constructors initialize the fields, but didn't know what to do about this block zeroing code.  Would you have to add it back with deterministic GC?

Reply:
> Yes, I'd like to make the constructors initialize the fields, but didn't know what to do about this block zeroing code. Would you have to add it back with deterministic CDS?

I am not sure yet. I think we may not need it because CDS doesn't copy HashtableEntries into the archive anymore.

Reply:
Instead of block zeroing hashtableEntry, you could add a field to zero in between _hash and _next and that would not have stray unaligned bytes.
template <MEMFLAGS F> class BasicHashtableEntry {
  friend class VMStructs;
private:
  unsigned int         _hash;           // 32-bit hash for item

  // Link to next element in the linked list for this bucket.
  BasicHashtableEntry<F>* _next;